datatypes/src/primitives/datetime.rs

```diff
@@ -0,0 +1,668 @@
+    }
+
+    pub fn parse_from_str(
+        input: &str,
+        format: &DateTimeParseFormat,
+    ) -> Result<Self, DateTimeError> {
+        match (format.has_time(), format.has_tz()) {
+            (true, true) => {
+                let datetime = chrono::DateTime::<chrono::FixedOffset>::parse_from_str(
+                    input,
+                    format._to_parse_format(),
+                )
+                .map_err(|e| DateTimeError::DateParse {
+                    source: Box::new(e),
+                })?;
+
+                Ok(Self {
+                    datetime: datetime.naive_utc(),
+                    offset: UtcOffset::_from_fixed_offset(datetime.offset().fix()),
+                })
+            }
+            (true, false) => {
+                let datetime =
+                    chrono::NaiveDateTime::parse_from_str(input, format._to_parse_format())
+                        .map_err(|e| DateTimeError::DateParse {
+                            source: Box::new(e),
+                        })?;
+
+                Ok(Self {
+                    datetime,
+                    offset: UTC,
+                })
+            }
+            (false, true) => Err(DateTimeError::CannotParseOnlyDateWithTimeZone),
+            (false, false) => {
+                let datetime = chrono::NaiveDate::parse_from_str(input, format._to_parse_format())
+                    .map_err(|e| DateTimeError::DateParse {
+                        source: Box::new(e),
+                    })?
+                    .and_hms(0, 0, 0);
+
+                Ok(Self {
+                    datetime,
+                    offset: UTC,
+                })
+            }
+        }
+    }
+
+    pub fn parse_from_rfc3339(input: &str) -> Result<Self, DateTimeError> {
+        let date_time = chrono::DateTime::<chrono::FixedOffset>::parse_from_rfc3339(input)
+            .map_err(|e| DateTimeError::DateParse {
+                source: Box::new(e),
+            })?;
+
+        Ok(date_time.into())
+    }
+
+    // TODO: is this inprecise?
+    pub fn parse_relaxed(input: &str) -> Result<Self, DateTimeError> {
+        let date_time = chrono::DateTime::<chrono::FixedOffset>::from_str(input).map_err(|e| {
+            DateTimeError::DateParse {
+                source: Box::new(e),
+            }
+        })?;
+
+        Ok(date_time.into())
+    }
+
+    pub fn format(&self, format: &DateTimeParseFormat) -> String {
+        let chrono_date_time: chrono::DateTime<chrono::FixedOffset> = self.into();
+        let parse_format = format._to_parse_format();
+
+        chrono_date_time.format(parse_format).to_string()
+    }
+
+    pub fn to_rfc3339(self) -> String {
+        let chrono_date_time: chrono::DateTime<chrono::FixedOffset> = self.into();
+        chrono_date_time.to_rfc3339()
+    }
+
+    /// Now in UTC.
+    pub fn now() -> Self {
+        chrono::offset::Utc::now().into()
+    }
+
+    pub fn day_of_year(&self) -> u16 {
+        // TODO: calculate offset?
```

Comment:
what offset?

Reply:
The reason I did not store a `DateTime<FixedOffset>` directly was for `chrono`, that `MIN` and `MAX` can no longer be `const`. But I guess for day of year the timezone matters?…

Moreover, maybe they don't have to be `const`… idk

Reply:
[6cc01e7](https://github.com/geo-engine/geoengine/pull/523/commits/6cc01e7e8c81f7a5e6eda1570bc2b3ab8fdecc0e)